<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.553">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>A Tutorial for Bayesian Integrative Factor Models - 2&nbsp; Case study: nutrition data</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./Gene.html" rel="next">
<link href="./Installation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./Nutrition.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Case study: nutrition data</span></a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">A Tutorial for Bayesian Integrative Factor Models</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Installation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Package Installation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Nutrition.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Case study: nutrition data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Gene.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Case study: gene expression data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#loading-and-previewing-the-data" id="toc-loading-and-previewing-the-data" class="nav-link active" data-scroll-target="#loading-and-previewing-the-data"><span class="header-section-number">2.1</span> Loading and previewing the data</a></li>
  <li><a href="#data-preprocessing" id="toc-data-preprocessing" class="nav-link" data-scroll-target="#data-preprocessing"><span class="header-section-number">2.2</span> Data preprocessing</a></li>
  <li><a href="#model-fitting" id="toc-model-fitting" class="nav-link" data-scroll-target="#model-fitting"><span class="header-section-number">2.3</span> Model fitting</a></li>
  <li><a href="#post-processing" id="toc-post-processing" class="nav-link" data-scroll-target="#post-processing"><span class="header-section-number">2.4</span> Post processing</a></li>
  <li><a href="#visualization" id="toc-visualization" class="nav-link" data-scroll-target="#visualization"><span class="header-section-number">2.5</span> Visualization</a></li>
  <li><a href="#mean-squared-error-mse" id="toc-mean-squared-error-mse" class="nav-link" data-scroll-target="#mean-squared-error-mse"><span class="header-section-number">2.6</span> Mean squared error (MSE)</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Case study: nutrition data</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="loading-and-previewing-the-data" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="loading-and-previewing-the-data"><span class="header-section-number">2.1</span> Loading and previewing the data</h2>
<p>The data used in this section is from a large multi-site study investigating health and diet among Hispanic/Latino adults in the United States <span class="citation" data-cites="de2022shared">(<a href="references.html#ref-de2022shared" role="doc-biblioref">De Vito et al. 2022</a>)</span>. This data is not publicly available. Please contact the authors of the original study for access.</p>
<p>However, you can use this simulated data instead: <a href="https://github.com/Mavis-Liang/Bayesian_integrative_FA_tutorial/raw/refs/heads/main/RDS/simulated_nutrition_data.rds">simulated_nutrition_data.rds</a> (if it does not triggers instant download, you can find it in the repo <a href="https://github.com/Mavis-Liang/Bayesian_integrative_FA_tutorial/blob/main/RDS/simulated_nutrition_data.rds">site</a>). To read in the data, use <code>readRDS()</code>. The loaded object contains data in both <code>Y_mat</code> and <code>Y_list</code> formats, as well as covariates and other information. It is generated with the Scenario 4 in the manuscript (see <a href="https://github.com/Mavis-Liang/Bayesian_integrative_FA_tutorial/blob/main/functions/gen_scenario4.R">code</a>). Note that instead of having 6 studies, this simulated data has 12 studies.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"./Data/dataLAT_projale2.rda"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The resulting object is a list of 6 data frames, each corresponding to a different study. Each data frame contains information about the nutritional intake of individuals, and the columns represent different nutrients. From Study 1 to Study 6, the number of individuals (<span class="math inline">\(N_s\)</span>) are 1364, 1517, 2210, 5184, 2478, and 959, respectively, and the number of nutrients (<span class="math inline">\(P\)</span>) are all 42.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check how many studies in the list</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(X_s2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Dimension of each study</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(X_s2, dim)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] 1364   42

[[2]]
[1] 1517   42

[[3]]
[1] 2210   42

[[4]]
[1] 5184   42

[[5]]
[1] 2478   42

[[6]]
[1] 959  42</code></pre>
</div>
</div>
<p>Let’s take a look at the first few rows of the first data frame to get an idea of the data structure.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>X_s2[[<span class="dv">1</span>]][<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Animal Protein (g) Vegetable Protein (g) Cholesterol (mg)  SCSFA MCSFA
1            28.9560               14.7440          256.761 0.2665 0.939
2            33.6675                8.9710          104.217 0.2180 0.520
3            70.0000               31.0635          207.902 0.9845 1.692
4            20.6700               13.8240          148.921 0.0625 0.239
5            15.4250               10.5550           65.060 0.0090 0.033</code></pre>
</div>
</div>
<p>We note that the data we have available is different from the original data (cite). The original data is a collection of 12 studies, and there are known covariates for each individuals, like the one we simulated in the previous section. However, for the purpose of this case study, the data we used are collapsed into 6 studies, and only the nutritional intake data are available.</p>
</section>
<section id="data-preprocessing" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="data-preprocessing"><span class="header-section-number">2.2</span> Data preprocessing</h2>
<p>Some individuals have missing values for all nutrients, thus we will remove these individuals from the data. Also, there are some nutrition intake are less than zero, for which we will replace with 0. We then apply a log transformation to the data.</p>
<p>We first count how many NA values and negative values are in each study.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>count_na_and_negatives <span class="ot">&lt;-</span> <span class="cf">function</span>(df) {</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Count NA values</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  na_count <span class="ot">&lt;-</span> <span class="fu">sum</span>(<span class="fu">is.na</span>(df))</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Count negative values</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  negative_count <span class="ot">&lt;-</span> <span class="fu">sum</span>(df <span class="sc">&lt;</span> <span class="dv">0</span>, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print counts</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Number of NAs:"</span>, na_count, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Number of negative values:"</span>, negative_count, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">lapply</span>(X_s2, count_na_and_negatives))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of NAs: 1344 
Number of negative values: 0 
Number of NAs: 1344 
Number of negative values: 1 
Number of NAs: 1344 
Number of negative values: 0 
Number of NAs: 1344 
Number of negative values: 2 
Number of NAs: 1344 
Number of negative values: 1 
Number of NAs: 1344 
Number of negative values: 0 </code></pre>
</div>
</div>
<p>We will define a function to process the data, which removes rows where all values are NA. We also define a function that replaces negative values with 0, and applies a log transformation to the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>process_study_data <span class="ot">&lt;-</span> <span class="cf">function</span>(df) {</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Remove rows where all values are NA</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  cleaned_df <span class="ot">&lt;-</span> df[<span class="sc">!</span><span class="fu">apply</span>(df, <span class="dv">1</span>, <span class="cf">function</span>(row) <span class="fu">all</span>(<span class="fu">is.na</span>(row))), , drop <span class="ot">=</span> <span class="cn">FALSE</span>]</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Count remaining rows</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  remaining_rows <span class="ot">&lt;-</span> <span class="fu">nrow</span>(cleaned_df)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Print results for the study</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cat</span>(<span class="st">"Remaining rows:"</span>, remaining_rows, <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(cleaned_df)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>Y_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(X_s2, process_study_data)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Remaining rows: 1332 
Remaining rows: 1485 
Remaining rows: 2178 
Remaining rows: 5152 
Remaining rows: 2446 
Remaining rows: 927 </code></pre>
</div>
</div>
<p>The numbers of individuals in each study left for analysis (<span class="math inline">\(N_s\)</span>) are 1332, 1485, 2178, 5152, 2446, and 927, respectively.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace negative values with 0, then log(x+0.01) + 0.01</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>replace_negatives <span class="ot">&lt;-</span> <span class="cf">function</span>(df) {</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Replace negative values with 0</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  df[df <span class="sc">&lt;</span> <span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Apply log transformation. Add 0.01 to avoid log(0).</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>  transformed_df <span class="ot">&lt;-</span> <span class="fu">log</span>(df <span class="sc">+</span> <span class="fl">0.01</span>)</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(transformed_df)</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>Y_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(Y_list, replace_negatives)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the processed data</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(<span class="fu">lapply</span>(Y_list, count_na_and_negatives))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Number of NAs: 0 
Number of negative values: 11910 
Number of NAs: 0 
Number of negative values: 11222 
Number of NAs: 0 
Number of negative values: 15006 
Number of NAs: 0 
Number of negative values: 36230 
Number of NAs: 0 
Number of negative values: 19349 
Number of NAs: 0 
Number of negative values: 6707 </code></pre>
</div>
</div>
<p>Now we don’t have any NA values or negative values in the data.</p>
<p>The assumptions for factor models require that each variable has a mean of 0. Therefore, for each study, we will center the data for each column. We note that for some model (Stack FA, Ind FA, BMSFA, and Tetris), this step is handeled internally, and for MOM-SS, the random intercepts are estimated, so we do not need to center the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>Y_list_scaled <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  Y_list, <span class="cf">function</span>(x) <span class="fu">scale</span>(x, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>Y_mat_scaled <span class="ot">&lt;-</span> Y_list_scaled <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="model-fitting" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="model-fitting"><span class="header-section-number">2.3</span> Model fitting</h2>
<p>We recommend running model fitting and post-processing in a high-performance computing environment, as the model fitting process can be computationally intensive. PFA and Tetris are particularly computationally expensive, where PFA requires more than 10 hours to run, and Tetris requires more than 4 days to run. Other models can be completed in half an hour. We recommend at least 5GB of memory for running the models and post-processing.</p>
<p>For each model, we over-specify the numbers of factors to .. [to be written].</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Stack FA</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>Y_mat <span class="ot">=</span>  Y_list <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>fit_stackFA <span class="ot">&lt;-</span> MSFA<span class="sc">::</span><span class="fu">sp_fa</span>(Y_mat, <span class="at">k =</span> <span class="dv">6</span>, <span class="at">scaling =</span> <span class="cn">FALSE</span>, <span class="at">centering =</span> <span class="cn">TRUE</span>, </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>                              <span class="at">control =</span> <span class="fu">list</span>(<span class="at">nrun =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">800</span>))</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Ind FA</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>fit_indFA <span class="ot">&lt;-</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>      <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>        j_s <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">8</span>, <span class="dv">8</span>, <span class="dv">8</span>, <span class="dv">8</span>, <span class="dv">8</span>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>        MSFA<span class="sc">::</span><span class="fu">sp_fa</span>(Y_list[[s]], <span class="at">k =</span> j_s[s], <span class="at">scaling =</span> <span class="cn">FALSE</span>, <span class="at">centering =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">control =</span> <span class="fu">list</span>(<span class="at">nrun =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">8000</span>))</span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>      })</span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a><span class="co"># PFA</span></span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>N_s <span class="ot">&lt;-</span> <span class="fu">sapply</span>(Y_list, nrow)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>fit_PFA <span class="ot">&lt;-</span> <span class="fu">PFA</span>(<span class="at">Y=</span><span class="fu">t</span>(Y_mat_scaled),</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>                        <span class="at">latentdim =</span> <span class="dv">6</span>,</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a>                        <span class="at">grpind =</span> <span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>,</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>                                    <span class="at">times =</span> N_s),</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>                <span class="at">Thin =</span> <span class="dv">5</span>,</span>
<span id="cb16-20"><a href="#cb16-20" aria-hidden="true" tabindex="-1"></a>                <span class="at">Total_itr =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">8000</span>)</span>
<span id="cb16-21"><a href="#cb16-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-22"><a href="#cb16-22" aria-hidden="true" tabindex="-1"></a><span class="co"># MOM-SS</span></span>
<span id="cb16-23"><a href="#cb16-23" aria-hidden="true" tabindex="-1"></a>Y_mat <span class="ot">=</span>  Y_list <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .) <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb16-24"><a href="#cb16-24" aria-hidden="true" tabindex="-1"></a><span class="co"># Construct the membership matrix</span></span>
<span id="cb16-25"><a href="#cb16-25" aria-hidden="true" tabindex="-1"></a>N_s <span class="ot">&lt;-</span> <span class="fu">sapply</span>(Y_list, nrow)</span>
<span id="cb16-26"><a href="#cb16-26" aria-hidden="true" tabindex="-1"></a>M_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb16-27"><a href="#cb16-27" aria-hidden="true" tabindex="-1"></a>   <span class="cf">for</span>(s <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>){</span>
<span id="cb16-28"><a href="#cb16-28" aria-hidden="true" tabindex="-1"></a>     M_list[[s]] <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="dv">1</span>, <span class="at">nrow =</span> N_s[s], <span class="at">ncol =</span> <span class="dv">1</span>)</span>
<span id="cb16-29"><a href="#cb16-29" aria-hidden="true" tabindex="-1"></a>   }</span>
<span id="cb16-30"><a href="#cb16-30" aria-hidden="true" tabindex="-1"></a>M <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(<span class="fu">bdiag</span>(M_list))</span>
<span id="cb16-31"><a href="#cb16-31" aria-hidden="true" tabindex="-1"></a>fit_MOMSS <span class="ot">&lt;-</span> BFR.BE<span class="sc">::</span><span class="fu">BFR.BE.EM.CV</span>(<span class="at">x =</span> Y_mat, <span class="at">v =</span> <span class="cn">NULL</span>, </span>
<span id="cb16-32"><a href="#cb16-32" aria-hidden="true" tabindex="-1"></a>                                  <span class="at">b =</span> M, <span class="at">q =</span> <span class="dv">6</span>, <span class="at">scaling =</span> <span class="cn">FALSE</span>)</span>
<span id="cb16-33"><a href="#cb16-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-34"><a href="#cb16-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-35"><a href="#cb16-35" aria-hidden="true" tabindex="-1"></a><span class="co"># SUFA</span></span>
<span id="cb16-36"><a href="#cb16-36" aria-hidden="true" tabindex="-1"></a>fit_SUFA <span class="ot">&lt;-</span> SUFA<span class="sc">::</span><span class="fu">fit_SUFA</span>(Y_list_scaled, <span class="at">qmax=</span><span class="dv">6</span>, <span class="at">nrun =</span> <span class="dv">10000</span>)</span>
<span id="cb16-37"><a href="#cb16-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-38"><a href="#cb16-38" aria-hidden="true" tabindex="-1"></a><span class="co"># BMSFA</span></span>
<span id="cb16-39"><a href="#cb16-39" aria-hidden="true" tabindex="-1"></a>fit_BMSFA <span class="ot">&lt;-</span> MSFA<span class="sc">::</span><span class="fu">sp_msfa</span>(Y_list, <span class="at">k =</span> <span class="dv">6</span>, <span class="at">j_s =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>),</span>
<span id="cb16-40"><a href="#cb16-40" aria-hidden="true" tabindex="-1"></a>                           <span class="at">outputlevel =</span> <span class="dv">1</span>, <span class="at">scaling =</span> <span class="cn">FALSE</span>, </span>
<span id="cb16-41"><a href="#cb16-41" aria-hidden="true" tabindex="-1"></a>                           <span class="at">centering =</span> <span class="cn">TRUE</span>,</span>
<span id="cb16-42"><a href="#cb16-42" aria-hidden="true" tabindex="-1"></a>                           <span class="at">control =</span> <span class="fu">list</span>(<span class="at">nrun =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">8000</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Fitting Tetris requires a 3-step process. First, we run <code>tetris()</code> to draw posterior samples of the model parameters, including <span class="math inline">\(\mathcal{T}\)</span>. Then we run <code>choose.A()</code> to choose the best <span class="math inline">\(\mathcal{T}\)</span> based on the posterior samples. Finally, we run <code>tetris()</code> again with the chosen <span class="math inline">\(\mathcal{T}\)</span> to obtain the final model. Hyperparameters <span class="math inline">\(\alpha_{\mathcal{T}}\)</span> are set to 1.25 times the number of studies.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tetris</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>set_alpha <span class="ot">&lt;-</span> <span class="fu">ceiling</span>(<span class="fl">1.25</span><span class="sc">*</span><span class="dv">6</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>fit_Tetris <span class="ot">&lt;-</span> <span class="fu">tetris</span>(Y_list, <span class="at">alpha=</span>set_alpha, <span class="at">beta=</span><span class="dv">1</span>, <span class="at">nprint =</span> <span class="dv">200</span>, </span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">nrun=</span><span class="dv">10000</span>, <span class="at">burn=</span><span class="dv">8000</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>big_T <span class="ot">&lt;-</span> <span class="fu">choose.A</span>(fit_Tetris, <span class="at">alpha_IBP=</span>set_alpha, <span class="at">S=</span><span class="dv">6</span>)</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>run_fixed <span class="ot">&lt;-</span> <span class="fu">tetris</span>(Y_list, <span class="at">alpha=</span>set_alpha, <span class="at">beta=</span><span class="dv">1</span>, </span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">fixed=</span><span class="cn">TRUE</span>, <span class="at">A_fixed=</span>big_T, <span class="at">nprint =</span> <span class="dv">200</span>, </span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">nrun=</span><span class="dv">10000</span>, <span class="at">burn=</span><span class="dv">8000</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="post-processing" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="post-processing"><span class="header-section-number">2.4</span> Post processing</h2>
<p>Post processing includes calculating the point estimates of the factor loadings and covariance matrix from the posterior samples, and determing the number of factors for each model.</p>
<p>For methods of MOM-SS, SUFA, and Tetris, the number of factors is determined internally in the algorithms, therefore, the output of the models are final results.</p>
<p>For MOM-SS, <span class="math inline">\(\Phi\)</span> is directly obtained with its post-processed common loadings in the fitted output. Th common covariance is calculated with <span class="math inline">\(\Phi\Phi^\top\)</span>. The marginal covariance matrix <span class="math inline">\(\Sigma_{\text{marginal}}\)</span> is calculated by adding the estimated study-specific error covariances to the common variance. The study-specific intercepts <span class="math inline">\(\alpha\)</span> and the coefficients for the known covariates <span class="math inline">\(B\)</span> are also extracted from the fitted output.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>post_MOMSS <span class="ot">&lt;-</span> <span class="cf">function</span>(fit, <span class="at">version =</span> <span class="dv">2</span>){ <span class="co"># version 1: M, version 2: Mpost</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>  est_Phi <span class="ot">&lt;-</span> fit<span class="sc">$</span>M</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (version<span class="sc">==</span><span class="dv">2</span>){est_Phi <span class="ot">&lt;-</span> fit<span class="sc">$</span>Mpost}</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  est_SigmaPhi <span class="ot">&lt;-</span> <span class="fu">tcrossprod</span>(est_Phi)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Marginal covariance</span></span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">dim</span>(fit<span class="sc">$</span>sigma)[<span class="dv">2</span>]</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>  est_PsiList <span class="ot">&lt;-</span> est_SigmaMarginal <span class="ot">&lt;-</span>  <span class="fu">list</span>()</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(s <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>S){</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    est_PsiList[[s]] <span class="ot">&lt;-</span> fit<span class="sc">$</span>sigma[,s]</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>    est_SigmaMarginal[[s]] <span class="ot">&lt;-</span> est_SigmaPhi <span class="sc">+</span> <span class="fu">diag</span>(fit<span class="sc">$</span>sigma[,s])</span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># last S columns of fit$Theta are the study-specific intercepts</span></span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>  est_alphas <span class="ot">&lt;-</span> fit<span class="sc">$</span>Theta[, (<span class="fu">dim</span>(fit<span class="sc">$</span>Theta)[<span class="dv">2</span>]<span class="sc">-</span>S<span class="sc">+</span><span class="dv">1</span>)<span class="sc">:</span><span class="fu">dim</span>(fit<span class="sc">$</span>Theta)[<span class="dv">2</span>]]</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>  <span class="co"># The rest are coeficients for the known covariates</span></span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>  est_B <span class="ot">&lt;-</span> fit<span class="sc">$</span>Theta[, <span class="dv">1</span><span class="sc">:</span>(<span class="fu">dim</span>(fit<span class="sc">$</span>Theta)[<span class="dv">2</span>]<span class="sc">-</span>S)]</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">Phi =</span> est_Phi, <span class="at">SigmaPhi =</span> est_SigmaPhi, <span class="at">Psi =</span> est_PsiList, <span class="at">alpha =</span> est_alphas, <span class="at">B =</span> est_B,</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaMarginal =</span> est_SigmaMarginal))</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb18-21"><a href="#cb18-21" aria-hidden="true" tabindex="-1"></a>res_MOMSS <span class="ot">&lt;-</span> <span class="fu">post_MOMSS</span>(fit_MOMSS)</span>
<span id="cb18-22"><a href="#cb18-22" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_MOMSS, <span class="st">"Data/Rnutrition_MOMSS.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For SUFA, the shared and study-specific loading matrices, as well as the common and marginal covariance can be conveniently obtained via the <code>lam.est.all()</code>, <code>SUFA_shared_covmat()</code> and <code>sufa_marginal_covs()</code> functions. Error covariance is obtained by taking averages of the “residuals” fitted output. And the study-specific covariance matrices are calculated by subtracting the common covariance from the marginal covariance. Note that in the definition of SUFA, common covariance is <span class="math inline">\(\Phi\Phi^\top + \Sigma\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>post_SUFA <span class="ot">&lt;-</span> <span class="cf">function</span>(fit){</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>  all <span class="ot">&lt;-</span> <span class="fu">dim</span>(fit<span class="sc">$</span>Lambda)[<span class="dv">3</span>]</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>  burnin <span class="ot">&lt;-</span> <span class="fu">floor</span>(all <span class="sc">*</span> <span class="fl">0.8</span>) <span class="co"># We will use the last 20% samples</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a>  <span class="co"># shared and study-specific loading matrices</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a>  loadings <span class="ot">&lt;-</span> <span class="fu">lam.est.all</span>(fit, <span class="at">burn =</span> burnin)</span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Obtain common covariance matrix and loading from fitting</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>  est_Phi <span class="ot">&lt;-</span> loadings<span class="sc">$</span>Shared</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a>  est_SigmaPhi <span class="ot">&lt;-</span> <span class="fu">SUFA_shared_covmat</span>(fit, <span class="at">burn =</span> burnin)</span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a>  est_Psi <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">colMeans</span>(fit<span class="sc">$</span>residuals))</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Study-specific loadings</span></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a>  est_LambdaList <span class="ot">&lt;-</span> loadings<span class="sc">$</span>Study_specific</span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Obtain study-specific covariance matrices</span></span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">length</span>(fit<span class="sc">$</span>A)</span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a>  marginal_cov <span class="ot">&lt;-</span> <span class="fu">sufa_marginal_covs</span>(fit, <span class="at">burn =</span> burnin)</span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a>  est_SigmaLambdaList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (s <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>S) {</span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a>    est_SigmaLambdaList[[s]] <span class="ot">&lt;-</span> marginal_cov[,,s] <span class="sc">-</span> est_SigmaPhi</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">SigmaPhi =</span> est_SigmaPhi, <span class="at">Phi =</span> est_Phi, </span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaLambdaList =</span> est_SigmaLambdaList,</span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a>              <span class="at">LambdaList =</span> est_LambdaList, </span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a>              <span class="at">Psi =</span> est_Psi,</span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaMarginal =</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s) marginal_cov[,,s])</span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>              ))</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>res_SUFA <span class="ot">&lt;-</span> <span class="fu">post_SUFA</span>(fit_SUFA)</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_SUFA, <span class="st">"Data/Rnutrition_SUFA.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For Tetris, the common loading matrix <span class="math inline">\(\Phi\)</span> can be obtained through the <code>getLambda()</code> function. The common covariance matrix is calculated as <span class="math inline">\(\Phi\Phi^\top\)</span>. The study-specific loading matrices <span class="math inline">\(\Lambda_s\)</span> are obtained by multiplying the common loading matrix with the study-specific matrices <span class="math inline">\(T_s - P\)</span>. The study-specific covariance matrices are calculated as <span class="math inline">\(\Lambda_s\Lambda_s^\top\)</span>. The marginal covariance matrix is calculated as <span class="math inline">\(\Lambda T \Lambda^\top + \Psi\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>post_Tetris <span class="ot">&lt;-</span> <span class="cf">function</span>(fit){</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Estimated common covariance</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  A <span class="ot">&lt;-</span> fit<span class="sc">$</span>A[[<span class="dv">1</span>]]</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  Lambda <span class="ot">&lt;-</span> <span class="fu">getLambda</span>(fit,A)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">dim</span>(A)[<span class="dv">1</span>]</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>  est_Phi <span class="ot">&lt;-</span> <span class="fu">as.matrix</span>(Lambda[,<span class="fu">colSums</span>(A)<span class="sc">==</span>S])</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>  est_SigmaPhi <span class="ot">&lt;-</span> <span class="fu">tcrossprod</span>(est_Phi)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Estimated study-specific covariance</span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>  P <span class="ot">=</span> <span class="fu">diag</span>((<span class="fu">colSums</span>(A) <span class="sc">==</span> S)<span class="sc">*</span><span class="dv">1</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>  T_s <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>  est_LambdaList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(s <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>S){</span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>    T_s[[s]] <span class="ot">&lt;-</span> <span class="fu">diag</span>(A[s,])</span>
<span id="cb20-14"><a href="#cb20-14" aria-hidden="true" tabindex="-1"></a>    Lambda_s <span class="ot">&lt;-</span> Lambda <span class="sc">%*%</span> (T_s[[s]] <span class="sc">-</span> P)</span>
<span id="cb20-15"><a href="#cb20-15" aria-hidden="true" tabindex="-1"></a>    Lambda_s <span class="ot">&lt;-</span> Lambda_s[,<span class="sc">-</span><span class="fu">which</span>(<span class="fu">colSums</span>(Lambda_s <span class="sc">==</span> <span class="dv">0</span>) <span class="sc">==</span> <span class="fu">nrow</span>(Lambda_s))]</span>
<span id="cb20-16"><a href="#cb20-16" aria-hidden="true" tabindex="-1"></a>    Lambda_s <span class="ot">&lt;-</span> <span class="fu">matrix</span>(Lambda_s, <span class="at">nrow=</span><span class="fu">nrow</span>(Lambda))</span>
<span id="cb20-17"><a href="#cb20-17" aria-hidden="true" tabindex="-1"></a>    est_LambdaList[[s]] <span class="ot">&lt;-</span> Lambda_s}</span>
<span id="cb20-18"><a href="#cb20-18" aria-hidden="true" tabindex="-1"></a>  est_SigmaLambdaList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s){</span>
<span id="cb20-19"><a href="#cb20-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tcrossprod</span>(est_LambdaList[[s]])})</span>
<span id="cb20-20"><a href="#cb20-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb20-21"><a href="#cb20-21" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Estimated marginal covariance</span></span>
<span id="cb20-22"><a href="#cb20-22" aria-hidden="true" tabindex="-1"></a>  Psi <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb20-23"><a href="#cb20-23" aria-hidden="true" tabindex="-1"></a>  est_SigmaMarginal <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s){</span>
<span id="cb20-24"><a href="#cb20-24" aria-hidden="true" tabindex="-1"></a>    Psi[[s]] <span class="ot">&lt;-</span> <span class="fu">diag</span>(<span class="fu">Reduce</span>(<span class="st">"+"</span>, fit<span class="sc">$</span>Psi[[s]])<span class="sc">/</span><span class="fu">length</span>(fit<span class="sc">$</span>Psi[[s]]))</span>
<span id="cb20-25"><a href="#cb20-25" aria-hidden="true" tabindex="-1"></a>    Sigma_s <span class="ot">&lt;-</span> Lambda <span class="sc">%*%</span> T_s[[s]] <span class="sc">%*%</span> <span class="fu">t</span>(Lambda) <span class="sc">+</span> Psi[[s]]</span>
<span id="cb20-26"><a href="#cb20-26" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb20-27"><a href="#cb20-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb20-28"><a href="#cb20-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">Phi =</span> est_Phi, <span class="at">SigmaPhi =</span> est_SigmaPhi,</span>
<span id="cb20-29"><a href="#cb20-29" aria-hidden="true" tabindex="-1"></a>              <span class="at">LambdaList =</span> est_LambdaList, <span class="at">SigmaLambdaList =</span> est_SigmaLambdaList,</span>
<span id="cb20-30"><a href="#cb20-30" aria-hidden="true" tabindex="-1"></a>              <span class="at">Psi =</span> Psi, <span class="at">T_s =</span> T_s,</span>
<span id="cb20-31"><a href="#cb20-31" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaMarginal =</span> est_SigmaMarginal))</span>
<span id="cb20-32"><a href="#cb20-32" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb20-33"><a href="#cb20-33" aria-hidden="true" tabindex="-1"></a>res_Tetris <span class="ot">&lt;-</span> <span class="fu">post_Tetris</span>(run_fixed)</span>
<span id="cb20-34"><a href="#cb20-34" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_Tetris, <span class="st">"Data/Rnutrition_Tetris.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For PFA, the post-processing is a little bit tricky. According to the original paper, the common loading matrix <span class="math inline">\(\Phi\)</span> can be just the average of the estimated loadings from the posterior samples. However, we still apply OP as in BMSFA to get the common loadings, and the loadings in PFA should be the product of the estimated loadings and the square root of the variance of the factors. Therefore, we carry out multiplication for <span class="math inline">\(\Phi\)</span> for each posterior sample, and then OP the sequence. We also carry out multiplication for <span class="math inline">\(Q_s\)</span>, <span class="math inline">\(\Sigma_{\Phi}\)</span>, <span class="math inline">\(\Sigma_{s}\)</span>, <span class="math inline">\(\Sigma_{\Lambda_s}\)</span>, and <span class="math inline">\(\Psi\)</span> as in the definition of PFA, and then take the average of the results.</p>
<p>After that, the number of factors is determined by counting the number of columns in <span class="math inline">\(\Phi\)</span> that have all loadings less than <span class="math inline">\(10^{-3}\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>post_PFA <span class="ot">&lt;-</span> <span class="cf">function</span>(fit){</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> <span class="fu">nrow</span>(fit<span class="sc">$</span>Loading[[<span class="dv">1</span>]])</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  k <span class="ot">&lt;-</span> <span class="fu">ncol</span>(fit<span class="sc">$</span>Loading[[<span class="dv">1</span>]])</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  npost <span class="ot">&lt;-</span> <span class="fu">length</span>(fit<span class="sc">$</span>Loading)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  Q_list <span class="ot">&lt;-</span> fit<span class="sc">$</span>Pertmat</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">dim</span>(Q_list[[<span class="dv">1</span>]])[<span class="dv">2</span>]</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>  posteriorPhis <span class="ot">&lt;-</span> <span class="fu">array</span>(<span class="dv">0</span>, <span class="at">dim =</span> <span class="fu">c</span>(p, k, npost))</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>  <span class="co">#---estimated common loadings and shared variance---</span></span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Element-wise multiplication</span></span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>npost){</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>    posteriorPhis[,,i] <span class="ot">&lt;-</span> fit<span class="sc">$</span>Loading[[i]] <span class="sc">%*%</span> <span class="fu">diag</span>(fit<span class="sc">$</span>Latentsigma[[i]])</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Varimax for common loadings</span></span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true" tabindex="-1"></a>  est_Phi <span class="ot">&lt;-</span> MSFA<span class="sc">::</span><span class="fu">sp_OP</span>(posteriorPhis, <span class="at">itermax =</span> <span class="dv">10</span>, <span class="at">trace =</span> <span class="cn">FALSE</span>)<span class="sc">$</span>Phi</span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Estimated shared covariance, study-specific covariance matrix, and SigmaMarginal</span></span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true" tabindex="-1"></a>  sharevar <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true" tabindex="-1"></a>  est_SigmaLambdaList <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true" tabindex="-1"></a>  est_SigmaMarginal <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true" tabindex="-1"></a>  est_Psi <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(s <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>S){ <span class="co"># Loop over each study</span></span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true" tabindex="-1"></a>  post_SigmaLambda_s <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true" tabindex="-1"></a>  post_SigmaMarginal_s <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true" tabindex="-1"></a>  Psi <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>npost){ <span class="co"># Loop over each posterior sample</span></span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true" tabindex="-1"></a>    sharevar[[i]] <span class="ot">&lt;-</span> fit<span class="sc">$</span>Loading[[i]]<span class="sc">%*%</span><span class="fu">diag</span>(fit<span class="sc">$</span>Latentsigma[[i]]<span class="sc">^</span><span class="dv">2</span>)<span class="sc">%*%</span><span class="fu">t</span>(fit<span class="sc">$</span>Loading[[i]]) <span class="sc">+</span> </span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true" tabindex="-1"></a>      <span class="fu">diag</span>(fit<span class="sc">$</span>Errorsigma[[i]]<span class="sc">^</span><span class="dv">2</span>) <span class="co"># Get the shared variance</span></span>
<span id="cb21-29"><a href="#cb21-29" aria-hidden="true" tabindex="-1"></a>    Q_temp_inv <span class="ot">&lt;-</span> <span class="fu">solve</span>(</span>
<span id="cb21-30"><a href="#cb21-30" aria-hidden="true" tabindex="-1"></a>      <span class="fu">matrix</span>(Q_list[[i]][, s], p, p)</span>
<span id="cb21-31"><a href="#cb21-31" aria-hidden="true" tabindex="-1"></a>      )</span>
<span id="cb21-32"><a href="#cb21-32" aria-hidden="true" tabindex="-1"></a>    post_SigmaMarginal_s[[i]] <span class="ot">&lt;-</span> Q_temp_inv<span class="sc">%*%</span>sharevar[[i]]<span class="sc">%*%</span><span class="fu">t</span>(Q_temp_inv)</span>
<span id="cb21-33"><a href="#cb21-33" aria-hidden="true" tabindex="-1"></a>    post_SigmaLambda_s[[i]] <span class="ot">&lt;-</span> post_SigmaMarginal_s[[i]] <span class="sc">-</span> sharevar[[i]]</span>
<span id="cb21-34"><a href="#cb21-34" aria-hidden="true" tabindex="-1"></a>    Psi[[i]] <span class="ot">&lt;-</span> <span class="fu">diag</span>(fit<span class="sc">$</span>Errorsigma[[i]]<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb21-35"><a href="#cb21-35" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-36"><a href="#cb21-36" aria-hidden="true" tabindex="-1"></a>  est_SigmaMarginal[[s]] <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, post_SigmaMarginal_s)<span class="sc">/</span><span class="fu">length</span>(post_SigmaMarginal_s)</span>
<span id="cb21-37"><a href="#cb21-37" aria-hidden="true" tabindex="-1"></a>  est_SigmaLambdaList[[s]] <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, post_SigmaLambda_s)<span class="sc">/</span><span class="fu">length</span>(post_SigmaLambda_s)</span>
<span id="cb21-38"><a href="#cb21-38" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb21-39"><a href="#cb21-39" aria-hidden="true" tabindex="-1"></a>  est_Psi <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, Psi)<span class="sc">/</span><span class="fu">length</span>(Psi)</span>
<span id="cb21-40"><a href="#cb21-40" aria-hidden="true" tabindex="-1"></a>  est_SigmaPhi <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, sharevar)<span class="sc">/</span><span class="fu">length</span>(sharevar)</span>
<span id="cb21-41"><a href="#cb21-41" aria-hidden="true" tabindex="-1"></a>  est_Q <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, Q_list)<span class="sc">/</span><span class="fu">length</span>(Q_list)</span>
<span id="cb21-42"><a href="#cb21-42" aria-hidden="true" tabindex="-1"></a>  est_Q_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s) <span class="fu">matrix</span>(est_Q[, s], p, p))</span>
<span id="cb21-43"><a href="#cb21-43" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb21-44"><a href="#cb21-44" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Return the results</span></span>
<span id="cb21-45"><a href="#cb21-45" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">Phi =</span> est_Phi, <span class="at">SigmaPhi =</span> est_SigmaPhi, <span class="at">Psi =</span> est_Psi, <span class="at">Q =</span> est_Q_list,</span>
<span id="cb21-46"><a href="#cb21-46" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaLambdaList =</span> est_SigmaLambdaList,</span>
<span id="cb21-47"><a href="#cb21-47" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaMarginal =</span> est_SigmaMarginal))</span>
<span id="cb21-48"><a href="#cb21-48" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb21-49"><a href="#cb21-49" aria-hidden="true" tabindex="-1"></a>res_PFA <span class="ot">&lt;-</span> <span class="fu">post_PFA</span>(fit_PFA)</span>
<span id="cb21-50"><a href="#cb21-50" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_PFA, <span class="st">"Data/Rnutrition_PFA.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># columns have all loadings less than 10^-3</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>fun_neibour <span class="ot">&lt;-</span> <span class="cf">function</span>(Phi, <span class="at">threshold =</span> <span class="fl">1e-3</span>) {</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">sum</span>(<span class="fu">apply</span>(Phi, <span class="dv">2</span>, <span class="cf">function</span>(x) {</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>      <span class="fu">sum</span>(<span class="fu">abs</span>(x) <span class="sc">&lt;=</span> threshold) <span class="sc">&lt;</span> <span class="fu">length</span>(x)}</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    ))</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="co"># The post_PFA(fit_PFA) object</span></span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>res_PFA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_PFA.rds"</span>)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>Phi_PFA <span class="ot">&lt;-</span> res_PFA<span class="sc">$</span>Phi</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a>K_PFA <span class="ot">&lt;-</span> <span class="fu">fun_neibour</span>(Phi_PFA)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>K_PFA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6</code></pre>
</div>
</div>
<p>The estimated <span class="math inline">\(K\)</span> for PFA is 6.</p>
<p>Next, we show how to process the output of Stack FA, Ind FA, and BMSFA. While point estimates for the loadings and covariances can be obtained in the similar way as in PFA, the number of factors is determined with eigen value decompositions of the covariance matrix. Once the numbers of factors are determined, we have to run the models again with the correct number of factors, and then extract the final results.</p>
<p>Therefore, for Stack FA, we first extract the point estimates of the <span class="math inline">\(\Phi\)</span> by applying OP to the posterior samples of the loadings. The common covariance matrix is calculated as <span class="math inline">\(\Phi\Phi^\top\)</span>. The marginal covariance matrix is calculated as the average of its posterior samples.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>post_stackFA <span class="ot">&lt;-</span> <span class="cf">function</span>(fit, S){</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  est_Phi <span class="ot">&lt;-</span> MSFA<span class="sc">::</span><span class="fu">sp_OP</span>(fit<span class="sc">$</span>Lambda, <span class="at">trace=</span><span class="cn">FALSE</span>)<span class="sc">$</span>Phi</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  est_SigmaPhi <span class="ot">&lt;-</span> <span class="fu">tcrossprod</span>(est_Phi)</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  est_SigmaMarginal <span class="ot">&lt;-</span>  <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">apply</span>(fit<span class="sc">$</span>Sigma, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), mean)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  Psi_chain <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(fit<span class="sc">$</span>Sigma)[<span class="dv">3</span>]){</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>    Psi_chain[[i]] <span class="ot">&lt;-</span> fit<span class="sc">$</span>Sigma[, , i] <span class="sc">-</span> <span class="fu">tcrossprod</span>(fit<span class="sc">$</span>Lambda[, , i])</span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a>  est_Psi <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, Psi_chain)<span class="sc">/</span><span class="fu">length</span>(Psi_chain)</span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">Phi =</span> est_Phi, <span class="at">SigmaPhi =</span> est_SigmaPhi, <span class="at">Psi =</span> est_Psi,</span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaMarginal =</span> est_SigmaMarginal))</span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>res_stackFA <span class="ot">&lt;-</span> <span class="fu">post_stackFA</span>(fit_stackFA, <span class="at">S=</span><span class="dv">6</span>)</span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_stackFA, <span class="st">"Data/Rnutrition_StackFA.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After that, we determine the number of factors by eigen value decomposition of the common covariance matrix. We then run the model again with the correct number of factors, and extract the final results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>fun_eigen <span class="ot">&lt;-</span> <span class="cf">function</span>(Sig_mean) {</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  val_eigen <span class="ot">&lt;-</span> <span class="fu">eigen</span>(Sig_mean)<span class="sc">$</span>values</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  prop_var <span class="ot">&lt;-</span> val_eigen<span class="sc">/</span><span class="fu">sum</span>(val_eigen)</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>  choose_K <span class="ot">&lt;-</span> <span class="fu">length</span>(<span class="fu">which</span>(prop_var <span class="sc">&gt;</span> <span class="fl">0.05</span>))</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(choose_K)</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>res_stackFA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_StackFA.rds"</span>)</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>SigmaPhi_StackFA <span class="ot">&lt;-</span> res_stackFA<span class="sc">$</span>SigmaPhi </span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>K_StackFA <span class="ot">&lt;-</span> <span class="fu">fun_eigen</span>(SigmaPhi_StackFA)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>K_StackFA</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
</div>
<p>The estimated <span class="math inline">\(K\)</span> for Stack FA is 4.</p>
<p>Then we re-run the model with the correct number of factors, and extract the final results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>fit_stackFA_2 <span class="ot">&lt;-</span> MSFA<span class="sc">::</span><span class="fu">sp_fa</span>(Y_mat_scaled, <span class="at">k =</span> K_StackFA, <span class="at">scaling =</span> <span class="cn">FALSE</span>, <span class="at">centering =</span> <span class="cn">TRUE</span>, </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>                           <span class="at">control =</span> <span class="fu">list</span>(<span class="at">nrun =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">8000</span>))</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>res_stackFA_2 <span class="ot">&lt;-</span> <span class="fu">post_stackFA</span>(fit_stackFA_2, <span class="at">S=</span><span class="dv">6</span>)</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_stackFA_2, <span class="st">"Data/Rnutrition_StackFA_2.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We repeat this process for Ind FA and BMSFA.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ind FA</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>post_indFA <span class="ot">&lt;-</span> <span class="cf">function</span>(fit){</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Estimated study-specific covariance and loading</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">length</span>(fit_list)</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  est_LambdaList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s){</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>    MSFA<span class="sc">::</span><span class="fu">sp_OP</span>(fit_list[[s]]<span class="sc">$</span>Lambda, <span class="at">trace=</span><span class="cn">FALSE</span>)<span class="sc">$</span>Phi</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a>  est_SigmaLambdaList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(est_LambdaList, <span class="cf">function</span>(x) <span class="fu">tcrossprod</span>(x))</span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Marginal covariance matrices</span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>  est_SigmaMarginal <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s) {</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>    fit <span class="ot">&lt;-</span> fit_list[[s]]</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">apply</span>(fit<span class="sc">$</span>Sigma, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), mean)</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-16"><a href="#cb28-16" aria-hidden="true" tabindex="-1"></a>  Psi <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb28-17"><a href="#cb28-17" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(s <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>S){</span>
<span id="cb28-18"><a href="#cb28-18" aria-hidden="true" tabindex="-1"></a>    Psi_chain <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb28-19"><a href="#cb28-19" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">dim</span>(fit_list[[<span class="dv">1</span>]]<span class="sc">$</span>Sigma)[<span class="dv">3</span>]){</span>
<span id="cb28-20"><a href="#cb28-20" aria-hidden="true" tabindex="-1"></a>      Psi_chain[[i]] <span class="ot">&lt;-</span> fit_list[[s]]<span class="sc">$</span>Sigma[, , i] <span class="sc">-</span> <span class="fu">tcrossprod</span>(fit_list[[s]]<span class="sc">$</span>Lambda[, , i])</span>
<span id="cb28-21"><a href="#cb28-21" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb28-22"><a href="#cb28-22" aria-hidden="true" tabindex="-1"></a>    Psi[[s]] <span class="ot">&lt;-</span> <span class="fu">Reduce</span>(<span class="st">'+'</span>, Psi_chain)<span class="sc">/</span><span class="fu">length</span>(Psi_chain)</span>
<span id="cb28-23"><a href="#cb28-23" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb28-24"><a href="#cb28-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-25"><a href="#cb28-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">LambdaList =</span> est_LambdaList, <span class="at">SigmaLambdaList =</span> est_SigmaLambdaList, <span class="at">Psi =</span> Psi,</span>
<span id="cb28-26"><a href="#cb28-26" aria-hidden="true" tabindex="-1"></a>              <span class="at">SigmaMarginal =</span> est_SigmaMarginal))</span>
<span id="cb28-27"><a href="#cb28-27" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-28"><a href="#cb28-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-29"><a href="#cb28-29" aria-hidden="true" tabindex="-1"></a>res_indFA <span class="ot">&lt;-</span> <span class="fu">post_indFA</span>(fit_indFA)</span>
<span id="cb28-30"><a href="#cb28-30" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_indFA, <span class="st">"Data/Rnutrition_IndFA.rds"</span>)</span>
<span id="cb28-31"><a href="#cb28-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-32"><a href="#cb28-32" aria-hidden="true" tabindex="-1"></a><span class="co"># BMSFA</span></span>
<span id="cb28-33"><a href="#cb28-33" aria-hidden="true" tabindex="-1"></a>post_BMSFA <span class="ot">&lt;-</span> <span class="cf">function</span>(fit){</span>
<span id="cb28-34"><a href="#cb28-34" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Common covariance matrix and loading</span></span>
<span id="cb28-35"><a href="#cb28-35" aria-hidden="true" tabindex="-1"></a>  est_Phi <span class="ot">&lt;-</span> <span class="fu">sp_OP</span>(fit<span class="sc">$</span>Phi, <span class="at">trace=</span><span class="cn">FALSE</span>)<span class="sc">$</span>Phi</span>
<span id="cb28-36"><a href="#cb28-36" aria-hidden="true" tabindex="-1"></a>  est_SigmaPhi <span class="ot">&lt;-</span> <span class="fu">tcrossprod</span>(est_Phi)</span>
<span id="cb28-37"><a href="#cb28-37" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-38"><a href="#cb28-38" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Study-specific covariance matrices and loadings</span></span>
<span id="cb28-39"><a href="#cb28-39" aria-hidden="true" tabindex="-1"></a>  est_LambdaList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(fit<span class="sc">$</span>Lambda, <span class="cf">function</span>(x) <span class="fu">sp_OP</span>(x, <span class="at">trace=</span><span class="cn">FALSE</span>)<span class="sc">$</span>Phi)</span>
<span id="cb28-40"><a href="#cb28-40" aria-hidden="true" tabindex="-1"></a>  est_SigmaLambdaList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(est_LambdaList, <span class="cf">function</span>(x) <span class="fu">tcrossprod</span>(x))</span>
<span id="cb28-41"><a href="#cb28-41" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-42"><a href="#cb28-42" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Marginal covariance matrices</span></span>
<span id="cb28-43"><a href="#cb28-43" aria-hidden="true" tabindex="-1"></a>  S <span class="ot">&lt;-</span> <span class="fu">length</span>(est_SigmaLambdaList)</span>
<span id="cb28-44"><a href="#cb28-44" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Get point estimate of each Psi_s</span></span>
<span id="cb28-45"><a href="#cb28-45" aria-hidden="true" tabindex="-1"></a>  est_PsiList <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s) {</span>
<span id="cb28-46"><a href="#cb28-46" aria-hidden="true" tabindex="-1"></a>    <span class="fu">apply</span>(fit<span class="sc">$</span>psi[[s]], <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>), mean)</span>
<span id="cb28-47"><a href="#cb28-47" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb28-48"><a href="#cb28-48" aria-hidden="true" tabindex="-1"></a>  est_margin_cov <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span>S, <span class="cf">function</span>(s) {</span>
<span id="cb28-49"><a href="#cb28-49" aria-hidden="true" tabindex="-1"></a>    est_SigmaPhi <span class="sc">+</span> est_SigmaLambdaList[[s]] <span class="sc">+</span> <span class="fu">diag</span>(est_PsiList[[s]] <span class="sc">%&gt;%</span> <span class="fu">as.vector</span>())</span>
<span id="cb28-50"><a href="#cb28-50" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb28-51"><a href="#cb28-51" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb28-52"><a href="#cb28-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="fu">list</span>(<span class="at">Phi =</span> est_Phi, <span class="at">SigmaPhi =</span> est_SigmaPhi,</span>
<span id="cb28-53"><a href="#cb28-53" aria-hidden="true" tabindex="-1"></a>         <span class="at">LambdaList =</span> est_LambdaList, <span class="at">SigmaLambdaList =</span> est_SigmaLambdaList,</span>
<span id="cb28-54"><a href="#cb28-54" aria-hidden="true" tabindex="-1"></a>         <span class="at">PsiList =</span> est_PsiList,</span>
<span id="cb28-55"><a href="#cb28-55" aria-hidden="true" tabindex="-1"></a>         <span class="at">SigmaMarginal =</span> est_margin_cov))</span>
<span id="cb28-56"><a href="#cb28-56" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb28-57"><a href="#cb28-57" aria-hidden="true" tabindex="-1"></a>res_BMSFA <span class="ot">&lt;-</span> <span class="fu">post_BMSFA</span>(fit_BMSFA)</span>
<span id="cb28-58"><a href="#cb28-58" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_BMSFA, <span class="st">"Data/Rnutrition_BMSFA.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>SigmaLambda_IndFA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_IndFA.rds"</span>)<span class="sc">$</span>SigmaLambda</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>Js_IndFA <span class="ot">&lt;-</span> <span class="fu">lapply</span>(SigmaLambda_IndFA, fun_eigen)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>Js_IndFA <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4 5 5 5 4 4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>SigmaPhi_BMSFA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_BMSFA.rds"</span>)<span class="sc">$</span>SigmaPhi</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>K_BMSFA <span class="ot">&lt;-</span> <span class="fu">fun_eigen</span>(SigmaPhi_BMSFA)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>SigmaLambda_BMSFA <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_BMSFA.rds"</span>)<span class="sc">$</span>SigmaLambda</span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>Js_BMSFA <span class="ot">&lt;-</span> <span class="fu">lapply</span>(SigmaLambda_BMSFA, fun_eigen)</span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a>K_BMSFA <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>Js_BMSFA <span class="sc">%&gt;%</span> <span class="fu">unlist</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 2 2 2 2 2 2</code></pre>
</div>
</div>
<p>The estimated <span class="math inline">\(J_s\)</span> for Ind FA are 4, 5, 5, 5, 4, and 4. The estimated <span class="math inline">\(K\)</span> for BMSFA is 4 and the estimated <span class="math inline">\(J_s\)</span> are 2, 2, 2, 2, 2, and 2</p>
<p>Then we re-run the models with the correct number of factors, and extract the final results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ind FA</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>fit_indFA_2 <span class="ot">&lt;-</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>    j_s <span class="ot">=</span> <span class="fu">c</span>(<span class="dv">4</span>, <span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">5</span>, <span class="dv">4</span>, <span class="dv">4</span>)</span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a>    MSFA<span class="sc">::</span><span class="fu">sp_fa</span>(Y_list[[s]], <span class="at">k =</span> j_s[s], <span class="at">scaling =</span> <span class="cn">FALSE</span>, <span class="at">centering =</span> <span class="cn">TRUE</span>,</span>
<span id="cb35-6"><a href="#cb35-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">control =</span> <span class="fu">list</span>(<span class="at">nrun =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">8000</span>))</span>
<span id="cb35-7"><a href="#cb35-7" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb35-8"><a href="#cb35-8" aria-hidden="true" tabindex="-1"></a>res_indFA_2 <span class="ot">&lt;-</span> <span class="fu">post_indFA</span>(fit_indFA_2)</span>
<span id="cb35-9"><a href="#cb35-9" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_indFA_2, <span class="st">"Data/Rnutrition_IndFA_2.rds"</span>)</span>
<span id="cb35-10"><a href="#cb35-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-11"><a href="#cb35-11" aria-hidden="true" tabindex="-1"></a><span class="co"># BMSFA</span></span>
<span id="cb35-12"><a href="#cb35-12" aria-hidden="true" tabindex="-1"></a>fit_BMSFA_2 <span class="ot">&lt;-</span> MSFA<span class="sc">::</span><span class="fu">sp_msfa</span>(Y_list, <span class="at">k =</span> <span class="dv">4</span>, <span class="at">j_s =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>, <span class="dv">2</span>),</span>
<span id="cb35-13"><a href="#cb35-13" aria-hidden="true" tabindex="-1"></a>                           <span class="at">outputlevel =</span> <span class="dv">1</span>, <span class="at">scaling =</span> <span class="cn">FALSE</span>, </span>
<span id="cb35-14"><a href="#cb35-14" aria-hidden="true" tabindex="-1"></a>                           <span class="at">centering =</span> <span class="cn">TRUE</span>,</span>
<span id="cb35-15"><a href="#cb35-15" aria-hidden="true" tabindex="-1"></a>                           <span class="at">control =</span> <span class="fu">list</span>(<span class="at">nrun =</span> <span class="dv">10000</span>, <span class="at">burn =</span> <span class="dv">8000</span>))</span>
<span id="cb35-16"><a href="#cb35-16" aria-hidden="true" tabindex="-1"></a>res_BMSFA_2 <span class="ot">&lt;-</span> <span class="fu">post_BMSFA</span>(fit_BMSFA_2)</span>
<span id="cb35-17"><a href="#cb35-17" aria-hidden="true" tabindex="-1"></a><span class="fu">saveRDS</span>(res_BMSFA_2, <span class="st">"Data/Rnutrition_BMSFA_2.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now the final results are obtained and you can get the saved files mentioned above from the <a href="https://github.com/Mavis-Liang/Bayesian_integrative_FA_benchmark/tree/main/RDS">GitHub repository</a>.</p>
</section>
<section id="visualization" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="visualization"><span class="header-section-number">2.5</span> Visualization</h2>
<p>We can make some heatmap for the loadings. Please see the figures in the paper.</p>
</section>
<section id="mean-squared-error-mse" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="mean-squared-error-mse"><span class="header-section-number">2.6</span> Mean squared error (MSE)</h2>
<p>We can assess the goodness-of-fit of the models by reconstructing data and calculating the reconstruction errors. With estimated loadings and error covariances, we can estimate the factor scores for a new observation, <span class="math inline">\(\widehat{\mathbf{f}}_{is,(new)}\)</span> and <span class="math inline">\(\widehat{\mathbf{l}}_{is,(new)}\)</span>, derived by adapting the Bartlett method. Then we can use the estimated factor scores, together with the estimated loadings, to reconstruct <span class="math inline">\(\widehat{\mathbf{y}}_{is,(new)}\)</span> and calculate the reconstruction error which represents the fit of the models. To be specific, suppose that we have the multivariate data of a new observation from a specific study, <span class="math inline">\(\mathbf{y}_{is,(new)}\)</span>, we can estimate its factor score and reconstruct its <span class="math inline">\(\widehat{\mathbf{y}}_{is,(new)}\)</span> in the following ways:</p>
<p>If we divide the whole data into training set and test set, we can obtain the between the true and estimated <span class="math inline">\(\mathbf{y}_{is,(new)}\)</span> in the test set for all individuals in all studies with <span class="math inline">\(\frac{1}{P\sum _s^S N_s}\sum_{s}^S\sum_{i}^{N_s}\sum_{p}^P(\widehat{y}_{isp,(new)}-y_{isp,(new)})^2\)</span>.</p>
<p>In the following, we divide the nutrition data into training set (70%) and test set (30%) and calculate the MSE of each model. Note that <span class="math inline">\(\mathbf{y}_{new}\)</span> should also be centered except for MOM-SS.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>train_ratio <span class="ot">&lt;-</span> <span class="fl">0.7</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>train_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>test_list <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (s <span class="cf">in</span> <span class="fu">seq_along</span>(Y_list)) {</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>  N_s <span class="ot">&lt;-</span> <span class="fu">nrow</span>(Y_list[[s]])  <span class="co"># Number of rows in the study</span></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>  train_indices <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span>N_s, <span class="at">size =</span> <span class="fu">floor</span>(train_ratio <span class="sc">*</span> N_s), <span class="at">replace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a>  train_list[[s]] <span class="ot">&lt;-</span> Y_list[[s]][train_indices, ]</span>
<span id="cb36-10"><a href="#cb36-10" aria-hidden="true" tabindex="-1"></a>  test_list[[s]] <span class="ot">&lt;-</span> Y_list[[s]][<span class="sc">-</span>train_indices, ]</span>
<span id="cb36-11"><a href="#cb36-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb36-12"><a href="#cb36-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Test data has to be centered</span></span>
<span id="cb36-13"><a href="#cb36-13" aria-hidden="true" tabindex="-1"></a>test_list <span class="ot">&lt;-</span> <span class="fu">lapply</span>(test_list, as.matrix)</span>
<span id="cb36-14"><a href="#cb36-14" aria-hidden="true" tabindex="-1"></a>test_list_scaled <span class="ot">&lt;-</span> <span class="fu">lapply</span>(</span>
<span id="cb36-15"><a href="#cb36-15" aria-hidden="true" tabindex="-1"></a>  test_list, <span class="cf">function</span>(x) <span class="fu">scale</span>(x, <span class="at">center =</span> <span class="cn">TRUE</span>, <span class="at">scale =</span> <span class="cn">FALSE</span>)</span>
<span id="cb36-16"><a href="#cb36-16" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We fit the models in the training data (on high-performance computing clusters), and then we load the fitted models and calculate the MSE on the test set for each model. The numbers of factors we input are determined by the previous results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>fit_StackFA_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_stackFA_train.rds"</span>)</span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a>fit_IndFA_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_IndFA_train.rds"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a>fit_PFA_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_PFA_train.rds"</span>)</span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a>fit_MOMSS_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_MOMSS_train.rds"</span>)</span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a>fit_SUFA_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_SUFA_train.rds"</span>)</span>
<span id="cb37-6"><a href="#cb37-6" aria-hidden="true" tabindex="-1"></a>fit_BMSFA_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_BMSFA_train.rds"</span>)</span>
<span id="cb37-7"><a href="#cb37-7" aria-hidden="true" tabindex="-1"></a>fit_Tetris_train <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"Data/Rnutrition_Tetris_train.rds"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We used the derived MSE function to calculate the MSE for each model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Stack FA</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>Phi <span class="ot">&lt;-</span> fit_StackFA_train<span class="sc">$</span>Phi</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> fit_StackFA_train<span class="sc">$</span>Psi</span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>mse_stackFA <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> test_list_scaled[[s]] <span class="sc">%*%</span> <span class="fu">solve</span>(Psi) <span class="sc">%*%</span> Phi <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(Phi) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi) <span class="sc">%*%</span> Phi))</span>
<span id="cb38-8"><a href="#cb38-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list_scaled[[s]] <span class="sc">-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(Phi), <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb38-9"><a href="#cb38-9" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb38-10"><a href="#cb38-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Ind FA</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>LambdaList <span class="ot">&lt;-</span> fit_IndFA_train<span class="sc">$</span>LambdaList</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> fit_IndFA_train<span class="sc">$</span>Psi</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>mse_IndFA <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> test_list_scaled[[s]] <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> LambdaList[[s]] <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(LambdaList[[s]]) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> LambdaList[[s]]))</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list_scaled[[s]] <span class="sc">-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(LambdaList[[s]]), <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PFA</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>Phi <span class="ot">&lt;-</span> fit_PFA_train<span class="sc">$</span>Phi</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> fit_PFA_train<span class="sc">$</span>Psi</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>Q_list <span class="ot">&lt;-</span> fit_PFA_train<span class="sc">$</span>Q</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>mse_PFA <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> test_list_scaled[[s]] <span class="sc">%*%</span> <span class="fu">t</span>(Q_list[[s]]) <span class="sc">%*%</span> Phi <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(Phi) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi) <span class="sc">%*%</span> Phi))</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a>    Y_est <span class="ot">&lt;-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(Phi) <span class="sc">%*%</span> <span class="fu">t</span>(<span class="fu">solve</span>(Q_list[[s]]))</span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list_scaled[[s]] <span class="sc">-</span> Y_est, <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MOM-SS</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>Phi <span class="ot">&lt;-</span> fit_MOMSS_train<span class="sc">$</span>Phi</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> fit_MOMSS_train<span class="sc">$</span>Psi <span class="sc">%&gt;%</span> <span class="fu">lapply</span>(diag)</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s) {</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  fit_MOMSS_train<span class="sc">$</span>alpha[,s]</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>mse_MOMSS <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(test_list[[s]], <span class="dv">1</span>, <span class="cf">function</span>(row) {row <span class="sc">-</span> alpha[[s]]})) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> Phi <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(Phi) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> Phi))</span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>    Y_est <span class="ot">&lt;-</span> <span class="fu">t</span>(<span class="fu">apply</span>(scores <span class="sc">%*%</span> <span class="fu">t</span>(Phi), <span class="dv">1</span>, <span class="cf">function</span>(row) {row <span class="sc">+</span> alpha[[s]]}))</span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list[[s]] <span class="sc">-</span> Y_est, <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># SUFA</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>Phi <span class="ot">&lt;-</span> fit_SUFA_train<span class="sc">$</span>Phi</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>LambdaList <span class="ot">&lt;-</span> fit_SUFA_train<span class="sc">$</span>LambdaList</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> fit_SUFA_train<span class="sc">$</span>Psi</span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a>mse_SUFA <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a>    Omega <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Phi, LambdaList[[s]])</span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> test_list_scaled[[s]] <span class="sc">%*%</span> <span class="fu">solve</span>(Psi) <span class="sc">%*%</span> Omega <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(Omega) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi) <span class="sc">%*%</span> Omega))</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list_scaled[[s]] <span class="sc">-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(Omega), <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># BMSFA</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>Phi <span class="ot">&lt;-</span> fit_BMSFA_train<span class="sc">$</span>Phi</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a>LambdaList <span class="ot">&lt;-</span> fit_BMSFA_train<span class="sc">$</span>LambdaList</span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> fit_BMSFA_train<span class="sc">$</span>PsiList <span class="sc">%&gt;%</span>  <span class="fu">lapply</span>(as.vector) <span class="sc">%&gt;%</span> <span class="fu">lapply</span>(diag)</span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>mse_BMSFA <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>    Omega <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Phi, LambdaList[[s]])</span>
<span id="cb43-8"><a href="#cb43-8" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> test_list_scaled[[s]] <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> Omega <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(Omega) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> Omega))</span>
<span id="cb43-9"><a href="#cb43-9" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb43-10"><a href="#cb43-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list_scaled[[s]] <span class="sc">-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(Omega), <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb43-11"><a href="#cb43-11" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb43-12"><a href="#cb43-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Tetris</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>Phi <span class="ot">&lt;-</span> fit_Tetris_train<span class="sc">$</span>Phi</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>LambdaList <span class="ot">&lt;-</span> fit_Tetris_train<span class="sc">$</span>LambdaList</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>Marginal <span class="ot">&lt;-</span> fit_Tetris_train<span class="sc">$</span>SigmaMarginal</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>SigmaPhi <span class="ot">&lt;-</span> fit_Tetris_train<span class="sc">$</span>SigmaPhi</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>SigmaLambdaList <span class="ot">&lt;-</span> fit_Tetris_train<span class="sc">$</span>SigmaLambdaList</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>Psi <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s) {</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>  Marginal[[s]] <span class="sc">-</span> SigmaPhi <span class="sc">-</span> SigmaLambdaList[[s]]</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>T_s_list <span class="ot">&lt;-</span> fit_Tetris_train<span class="sc">$</span>T_s</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>mse_Tetris <span class="ot">&lt;-</span> <span class="dv">1</span><span class="sc">/</span>(<span class="dv">42</span> <span class="sc">*</span> <span class="fu">sum</span>(<span class="fu">sapply</span>(test_list, nrow)))<span class="sc">*</span><span class="fu">sum</span>(</span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>, <span class="cf">function</span>(s){</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>    Omega <span class="ot">&lt;-</span> <span class="fu">cbind</span>(Phi, LambdaList[[s]])</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>    scores <span class="ot">&lt;-</span> test_list_scaled[[s]] <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> Omega <span class="sc">%*%</span> mnormt<span class="sc">::</span><span class="fu">pd.solve</span>(<span class="fu">signif</span>(<span class="fu">t</span>(Omega) <span class="sc">%*%</span> <span class="fu">solve</span>(Psi[[s]]) <span class="sc">%*%</span> Omega))</span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>    <span class="fu">norm</span>(test_list_scaled[[s]] <span class="sc">-</span> scores <span class="sc">%*%</span> <span class="fu">t</span>(Omega), <span class="st">"F"</span>)<span class="sc">^</span><span class="dv">2</span></span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a>  })</span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We display the MSE of each model.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Stack FA: "</span>, mse_stackFA <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Ind FA: "</span>, mse_IndFA <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"PFA: "</span>, mse_PFA <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"MOM-SS: "</span>, mse_MOMSS <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"SUFA: "</span>, mse_SUFA <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"BMSFA: "</span>, mse_BMSFA <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">"Tetris: "</span>, mse_Tetris <span class="sc">%&gt;%</span> <span class="fu">round</span>(<span class="dv">3</span>)))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Stack FA: 0.482"
[1] "Ind FA: 0.462"
[1] "PFA: 0.672"
[1] "MOM-SS: 0.455"
[1] "SUFA: 0.421"
[1] "BMSFA: 0.437"
[1] "Tetris: 0.266"</code></pre>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-de2022shared" class="csl-entry" role="listitem">
De Vito, Roberta, Briana Stephenson, Daniela Sotres-Alvarez, Anna-Maria Siega-Riz, Josiemer Mattei, Maria Parpinel, Brandilyn A Peters, et al. 2022. <span>“Shared and Ethnic Background Site-Specific Dietary Patterns in the Hispanic Community Health Study/Study of Latinos (HCHS/SOL).”</span> <em>medRxiv</em>, 2022–06.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./Installation.html" class="pagination-link" aria-label="Package Installation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Package Installation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./Gene.html" class="pagination-link" aria-label="Case study: gene expression data">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Case study: gene expression data</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>